src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -40,12 +40,63 @@ internal Task<ManagedIdentityResponse> SendTokenRequestForManagedIdentityAsync(A
         // This method tries to create managed identity source for different sources, if none is created then defaults to IMDS.
         private static AbstractManagedIdentity SelectManagedIdentitySource(RequestContext requestContext)
         {
-            return 
-                ServiceFabricManagedIdentitySource.TryCreate(requestContext) ??
-                AppServiceManagedIdentitySource.TryCreate(requestContext) ?? 
-                CloudShellManagedIdentitySource.TryCreate(requestContext) ??
-                AzureArcManagedIdentitySource.TryCreate(requestContext) ??
-                new ImdsManagedIdentitySource(requestContext);
+            ManagedIdentitySource managedIdentitySource = GetManagedIdentitySource();
+
+            switch (managedIdentitySource)
+            {
+                case ManagedIdentitySource.ServiceFabric:
```

Comment:
Would be nice to move the source detection logic here and cache it, so we do not detect the source again.

Reply:
+1